src/tir/schedule/concrete_schedule.cc

```diff
@@ -208,6 +211,25 @@ Schedule ConcreteScheduleNode::Copy() const {
   }
 
 /******** Schedule: Schedule: Sampling ********/
+
+void ConcreteScheduleNode::Seed(support::LinearCongruentialEngine::TRandState seed) {
+  support::LinearCongruentialEngine(&rand_state_).Seed(seed == -1 ? std::random_device()() : seed);
+}
+support::LinearCongruentialEngine::TRandState ConcreteScheduleNode::ForkSeed() {
+  // In order for reproducibility, we computer the new seed using RNG's random state and a different
+  // set of parameters. Note that both 32767 and 1999999973 are prime numbers.
+  return (support::LinearCongruentialEngine(&rand_state_)() * 32767) % 1999999973;
+}
```

Comment:
It seems like ForkSeed is analogous to what is called "splitting" in the random number generator literature. I'm not quite an expert on this, but I did do a bit of research into PRNGS for the Threefry implementation we have. Everything I read says that there are no proofs to the validity of splitting LCGs (is the method you use here from a paper?). The paper ["Splittable Pseudorandom Number Generators using Cryptographic Hashing"](https://publications.lib.chalmers.se/records/fulltext/183348/local_183348.pdf) provides some good explanations.

In practice, I expect we will see some issues. If this function somehow perfectly bisects the space of random numbers generated by this PRNG, then we could expect to start seeing repeats of previous random numbers after 31 splits. Given that this splitting does not perfectly bisect the space, I'd assume that we start seeing repeats much sooner. Repeating portions of the search space may mean that we may no be able to visit the entire search space during tuning or that we may bias results towards a certain section of the space.

I'd suggest we adopt a splittable PRNG here as that appears the be what we need. Maybe we can find an existing implementation online as implementing your own PRNG can have subtle issues.

Reply:
LCGs are pretty easy to be cracked in terms of security.

To be clear, we don't split the RNG too many times. It is only used in terms of multi-threaded search where we split the RNG for each thread, where in practice we didn't see repetition or any problem caused by repetition when running tens of works.

Reply:
Don't most modern machines have at least 31 hyper threads, i.e we will split at least 31 times on those machines?

Reply:
Here is a quick example, supposing we have 64 threads and repeated for 10k times: https://gist.github.com/junrushao1994/ea986add81b01b89fd99a5a7d41d087a, and the result is that there is no repetition.

I actually agree with Tristan's theory in general, but in our particular case it is not a big problem. To make sure it works generally, we can switch PRNGs by replacing support::LCG with other random number generators :-)

Reply:
Coming late to the discussion. I read the thread yesterday evening and wanted
to spend more time thinking so did not reply immediately. 

Please let me try to summarize a few points from the current discussion:

As we generate random numbers, PRNG state circles through the space of states
and eventually will come back to its initial state to form a circle.
When we split the state of PRNG to start parallel generators, if the 
splitted states are "too close" in terms of their traversal distance,
then we get repeatitions, or streams that are correlated to each other.

The following pts about the PRNGs

- A0: When we split a PRNG in an adhoc way, there is a "possibility"
      that repeats can happen in different streams. The nature of such
      possibility depends on the splitting method, the seed and PRNG itself.
- A1: There is currently no proof of splitting LCG in general.


Besides the above two pts, we also need to consider the implication of 
using a possibly correlated PRNG streams. In our case, we use PRNG
to generate explorations in the search space and there are two possible tasks:
- T0: explore the space and find possible maxima in the space

To make things simple, let us assume that there are two streams in the 32 threads
that are exactly identical to each other. In this case, we waste the computation
from one of the thread, because it will result in exactly the same set of samples.
Because our goal is to explore the space and find a maximum point. This would mean
that the computation from that particular thread is wasted, and we get less statistical
efficiency from the sampling process.

At a high level, we have two parameters we can tweak, the number of sampling steps n,
and number of threads K. What we are saying is that the statistical efficiency of running
uniform sampling over the search space scales linearly with n, but perhaps less optimally
wrt to K. For other more complicated tasks, such as estimating density of certain regions.
We know that the sampling and averaging over time(n) always can give us the right estimation.
Correlation across streams, would make averaging over streams(K) less efficient because the
random numbers are not independent, but we will still get the correct mean as we increase n.

So in summary:

- A2: The impact of possibly correlated streams means that we could get less than
      full-linear efficiency in terms of number of thread K. The sampling will still    
      effective samples wrt to n(the number of samples we take in each thread).
       
Note that the real sampling scenario is much more complicated. As junru's experiments
showed that repeatition did not happen on the particular case for quite long period of time.
Additionally, the sampling steps are dependent on each other(they form a Markov Chain), 
so it is hard to tell the end effect correlation between random sequence (abcde) and (bcdef),
even if they are one step apart. While in most of the cases they can be many steps apart
emperically. To summarize


- A3: In this particular usecase, emperically we get non-repetitively sequences among the streams.
      This does not preclude that correlation won't happen(as it is not a proof), it does suggest
      that correlation may not be large in most cases.

The end effect of A2 has a quite close analogy in parallel computing: as we start to use
K thread, we may not exactly get Kx speedups. Except that in this case it is not due to
hardware reasons, it is due to the possible correlation. As in parallel computing, 
we can run the program longer, in our case increase n to compensate the possible loss
of efficiency. In short, A2 might suggest that parallel stream correlation may not be the problem
that we need to perfectly solve, as long as it does not become very bad(e.g. all streams are the same).

Yesterday I did not think of A2 in particular, which might change our perspective. So I would
like to share this summary here. Would be great to get your takes as well.

Reply:
@tqchen I happened to implement a BSGS discrete logarithm this morning. This is a simple but effective algorithm we use in high school competitive programming: https://gist.github.com/junrushao1994/d32f265f5b4815d4b346d6022e95f394.

I use this script to find out what the minimal number of trials is required for a first repeat to happen given `num_threads` threads, and here is the outcome when I set `num_threads=1000`.

```
k = 0, forked_seed = 1, repeat = 0
k = 1, forked_seed = 32767, repeat = 477805293
k = 2, forked_seed = 1073676289, repeat = 955610586
k = 3, forked_seed = 1151436593, repeat = 792763173
k = 4, forked_seed = 1123352159, repeat = 1465074278
k = 5, forked_seed = 880690861, repeat = 1324276493
k = 6, forked_seed = 1597831943, repeat = 1242547212
k = 7, forked_seed = 159983087, repeat = 1577951775
k = 8, forked_seed = 165882496, repeat = 12097717
k = 9, forked_seed = 1471819791, repeat = 670441922
k = 10, forked_seed = 1119742748, repeat = 130415288
k = 11, forked_seed = 611119031, repeat = 268696235
k = 12, forked_seed = 537559101, repeat = 53491731
k = 13, forked_seed = 199300256, repeat = 1422308282
k = 14, forked_seed = 471576507, repeat = 535435921
k = 15, forked_seed = 147613471, repeat = 482454183
k = 16, forked_seed = 850669543, repeat = 1171123831
k = 17, forked_seed = 1889291753, repeat = 1518153595
k = 18, forked_seed = 423706282, repeat = 325709943
k = 19, forked_seed = 1583929701, repeat = 1523336904
k = 20, forked_seed = 625213317, repeat = 2097645560
...
k = 990, forked_seed = 465632523, repeat = 1395274874
k = 991, forked_seed = 1381087097, repeat = 1524683345
k = 992, forked_seed = 81518328, repeat = 874365972
k = 993, forked_seed = 1111089621, repeat = 464689348
k = 994, forked_seed = 1074102788, repeat = 1779776079
k = 995, forked_seed = 1126529515, repeat = 113162479
k = 996, forked_seed = 993116317, repeat = 711897275
k = 997, forked_seed = 1442798429, repeat = 285912163
k = 998, forked_seed = 176761269, repeat = 918045815
k = 999, forked_seed = 1936579488, repeat = 43150205
min repeat: 1407035
```

In a word, in practice the repeat won't happen after 1407035 trials given the `forked_seed` within 1000 threads.

Reply:
@tqchen @junrushao1994  You both lay out a lot of interesting points here, but I'm not sure I have the expertise to evaluate them. The PRNGS themselves might appear simple, but analysis of their randomness is complicated and non-intuitive. Looking at the paper I linked above, you can get subtle bugs if the PRNG is used incorrectly. I've tested the LCG implemented in TVM with some PRNG test suites (you can try it yourself here: https://github.com/tkonolige/prng-tests), and it fails all of them. This result is unsurprising because LCGs aren't particularly good random number generators, but it just adds a little to my concern.

Given that we want to avoid any potential issues, why don't we just do things the right way and use a splittable PRNG? This page (https://www.pcg-random.org/posts/some-prng-implementations.html) lists some implementations of PRNGs including SplitMix which is splittable. (pcg-random appears to be a reputable source, it is run by the create of the PCG family of PRNGS). It seems like there is basically no overhead to just dropping this SplitMix implementation into the codebase. And then we won't have to worry about any bugs due to bad randomness.

Reply:
I don't want to block this PR on this, so I'm going to approve. But I would like us to fix this in the future.